src/csharp/Microsoft.Spark/Utils/PythonSerDe.cs

```diff
@@ -38,5 +39,18 @@ internal static object[] GetUnpickledObjects(Stream s)
             Debug.Assert(unpickledItems != null);
             return (unpickledItems as object[]);
         }
+
+        /// <summary>
+        /// Unpickles objects from memory buffer
+        /// </summary>
+        /// <param name="buffer">Pickled byte buffer</param>
+        /// <returns>Unpicked objects</returns>
+        internal static object[] GetUnpickledObjects(ReadOnlyMemory<byte> buffer)
+        {
+            var unpickler = new Unpickler();
+            var unpickledItems = unpickler.loads(buffer, stackCapacity: 100); // 100 because spark typically sends batches of 100 rows
```

Comment:
Should the `100` be configurable at all?  I know that for the Arrow format, the number of rows in a batch is configurable.

Reply:
@imback82 is the number of rows in a batch from spark configurable when using Pickling?

I have used 100 because this is what I have seen in the profiler. Also, this is only the initial capacity, not max capacity.

Reply:
100 is hard coded on the Spark side: https://github.com/apache/spark/blob/d9bcacf94b93fe76542b5c1fd852559075ef6faa/sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala#L77

Would it hurt the performance a lot if we just use default (4)?

Reply:
@imback82 thanks for the confirmation. I have set the capacity to avoid unnecessary copies when unpickling and I would prefer to keep it that way.

If we want to unpickle a list of 100 doubles (or any other type) currently an array of 4 elements is allocated, then 8 and 4 are copied, then 16 and 8 are copied, then 32 and 16 are copied then 64 and 32 are copied then 128 and 64 are copied. 

So we allocate 6 arrays of 252 elements in total and copy 124 elements. Setting the capacity to 100 allows us to allocate the array only once and perform 0 copies.

Reply:
@imback82, what situation are you optimizing for here?  Doesn't a stack of 100 suggest ~100 columns?  I'd expect this change would actually be a deoptimization for the common case where there are only a few columns.  Am I misunderstanding?

Reply:
Ah, I was misremembering the implemetation.  Looks like the stack ends up being used to also hold the deserialized row results.  Ok.

Reply:
@stephentoub  The unpickler pushes every deserialized object to an internal stack. It is used to support tuples, list, arrays and more.

Let's consider the following example: we want to serialize a List (spark always excepts a list as a result, it throws for arrays) of arrays of two doubles each:

```
var toSerialize = new List<object>()
{
   new object[2] { 1.0, 2.0 },
   new object[2] { 3.0, 4.0 },
}
```

The pickling format represents it as:

```
Opcodes.EMPTY_LIST (byte marker)
Opcodes.MARK
Opcodes.BINFLOAT
1.0 (8 bytes Int64 BigEndianInteger)
Opcodes.BINFLOAT
2.0
Opcodes.TUPLE2
Opcodes.BINFLOAT
3.0
Opcodes.BINFLOAT
4.0
Opcodes.TUPLE2
Opcodes.APPENDS
```

Now when it deserializes the provided input:

- it pushes everything to the mentioned stack
- when it encounters TUPLE2 mark it pops the last two items and allocates an array of two elements and stores poped items in it. It also pushes the array to the stack.
- when it encounters APPENDS mark it pops everything from the stack until it hits MARK, stores it in a list, reverts it (it's a stack) and then returns the list


Adding the items to the stack and resizing it was visible in the profiles I collected after optimizing the way we read/write from/to an array so I have introduced this new parameter to provide a hint to the unpickler and save few % when deserializing.

Reply:
Right, that's what I meant by "Looks like the stack ends up being used to also hold the deserialized row results." :wink:

Reply:
Given that, isn't 100 generally too small?  When processing the last of 100 rows, won't we already be ~100 deep, and we'll grow the array to handle the processing of the last row?

Reply:
> Right, that's what I meant by

I got an email with your first comment, started writing the response, posted it and then it turned out that in the meantime you wrote another one saying that you don't need the explanation anymore ;P

I hope that others will find it useful ;)

Reply:
> Given that, isn't 100 generally too small?

After setting it to 100 it got much smaller in the profiles. I was thinking about bigger number, but was not sure. Most benchmarks have batches of arrays of 100 items, some have of 100 items where every item is a tuple (so we would need 100).

Reply:
> some have of 100 items where every item is a tuple (so we would need 100

With 100 rows of just one double per row, doesn't the pickling stack actually need to hold 101 elements?

And with tuples, when processing the last tuple, wouldn't the reading of the last element push all of the elements of the tuple onto the stack, such that we'd grow beyond 100?

I'm not convinced any of this really matters, but if we're going to play games with the initial size, we should make sure it's big enough for the common cases to avoid whatever overheads we're trying to avoid.  I would suggest running all of the various TPCH benchmarks that are relevant and keeping track of the high-water mark on the stack size.

Reply:
You are right. The stack size should be... 102 (I forgot about the markers). I have verified that it's the perfect number for Q1, Q13 and Q8.